Conditionally disable fastprogress.progressbar in sample_posterior_predictive

rpgoldman opened the pull request:
Instead of creating the progress bar and not displaying it, don't create it at all.

This allows users to avoid some display issues in console mode.

I did this before for sample, but missed its use here.

michaelosthege reviewed (changes requested):
Hold on, the tests didn't catch it:

If `progressbar=False` and there are divergences, it blows up in line 634.

rpgoldman commented:
Thanks, @michaelosthege , I see the problem -- we need to make sure that the progress bar is there before writing to its comment field.

Here's a question -- is there a good way I could create a test for this? Do you have a simple model with a divergence I could put in the tests?

rpgoldman committed:
Conditionally disable progressbar.

If the progressbar is turned off, don't create it (instead of creating it and not displaying it).

michaelosthege commented:
`_pbar_data` is always defined, so you can drop line 622 and the corresponding condition in 633.

The model from @twiecki's blog post about funnels may give divergences.

rpgoldman commented:
> `_pbar_data` is always defined, so you can drop line 622 and the corresponding condition in 633.

Thanks for the simplifications.  I applied them, and then did a pass of style cleanups (pylint, black, pydocstyle).

rpgoldman commented:
@michaelosthege Would you please have one more look at this, and squash merge if it's ok? I think I have made all the changes you suggested, and done some clean-up, as well.

michaelosthege commented on pymc3/sampling.py:
```diff
@@ -226,7 +230,7 @@ def sample(
     **kwargs
 ):
-    """Draw samples from the posterior using the given step methods.
+    r"""Draw samples from the posterior using the given step methods.
```
Why this `r`-string? Does it have to do with doc formatting?

michaelosthege commented on pymc3/sampling.py:
```diff
@@ -823,7 +862,7 @@ def __exit__(self, exc_type, exc_val, exc_tb):
     @staticmethod
     def _run_slave(c, stepper_dumps, slave_end):
-        """Started on a separate process to perform stepping of a chain.
+        """Start a separate process to perform stepping of a chain.
```
No, this is wrong.
Should be: `This method is started on a separate process...`

michaelosthege reviewed (changes requested):
Did you use black for the formatting?
In my opinion there are a lot of unnecessary line breaks - probably due to a very conservative line length.

I'll go and edit a few strings.

michaelosthege commented:
The target coverage check failed, probably because line breaks in warnings increased the total number while not being covered by tests.

As much as I dislike vertically stretched out code, I also think that it's now technically fine.
So if everybody agrees we can merge even though it decreases coverage.

rpgoldman commented:
The vertically stretched out code, FWIW, isn't really me: it's formatting by black, which I believe to be the project standard.

Thanks for so carefully reviewing this.

michaelosthege commented:
Our `.pylintrc` already specifies a maximum line length of 100, which would allow most of these changes to be on one line.

How did you run the `black` command? Maybe we need to specify some settings somewhere (I'm not familiar with it yet).

michaelosthege commented:
From [blacks README](https://github.com/psf/black/blob/master/README.md), this should do the trick:

`black --line-length 100`

Their default is 88, so...

rpgoldman commented:
@michaelosthege When you get a chance, will you please look at the results of my re-running black with line length of 100?

I think this fixes the things you didn't like. Please let me know, and I will get this tidied up a little (I will try to separate the reformatting from the progress bar disabling) and merged.

rpgoldman committed:
Code formatting, simplification, and type hints.

@michaelostage found two simplifications that could me made in `_sample()`, added some type hints in passing, and fixed the formatting of the sample() docstring.
Pylint fixes, black reformatting and (partial) pydocstyle corrections.
fix docstring to accurately describe function

This version uses 100 as the line length, breaking up fewer warnings, etc.
Fix docstring for sample()

With help from @michaelostage, made some simplifications, cleaned up the source and docstring formatting, and added some type hints.
